File: 200/289 - Game of Life.c
void gameOfLife(int** board, int boardRowSize, int boardColSize) {
    int r, c, i, x, y, count;
    int r_range[] = {-1, -1, -1, 0, 0, 1, 1, 1};
    int c_range[] = {-1, 0, 1, -1, 1, -1, 0, 1};
    
    if (! (boardRowSize && boardColSize))
        return;
    
    for (r = 0; r < boardRowSize; ++r) {
        for (c = 0; c < boardColSize; ++c) {
            count = 0;
            
            for (i = 0; i < 8; ++i) {
                x = r + r_range[i];
                y = c + c_range[i];
                
                if (x >= 0 && x < boardRowSize && y >= 0 && y < boardColSize && (board[x][y] == 1 || board[x][y] == 2))
                    ++count;
            }
            
            if (board[r][c] && (count < 2 || count > 3))
                board[r][c] = 2;
            else if (count == 3 && ! board[r][c])
                board[r][c] = 3;
        }
    }
    
    for (r = 0; r < boardRowSize; ++r) {
        for (c = 0; c < boardColSize; ++c)
            board[r][c] %= 2;
    }
}

File: csharp/289.game-of-life.cs
/*
 * @lc app=leetcode id=289 lang=csharp
 *
 * [289] Game of Life
 */

using System.Linq;
using System;

public class Solution {

    private static (int, int)[] Dirs = new (int, int)[] {
        (-1, -1), (-1, 0), (-1, 1), (0, -1), (0, 1), (1, -1), (1, 0), (1, 1),
    };

    public void GameOfLife(int[][] board) {
        int m = board.Length, n = board[0].Length;
        Func<(int, int), int> Sum = t => (t.Item1 >= 0 && t.Item1 < m && t.Item2 >= 0 && t.Item2 < n && (board[t.Item1][t.Item2] == 1 || board[t.Item1][t.Item2] == 2)) ? 1 : 0;
        for (int i = 0; i < m; ++i) {
            for (int j = 0; j < n; ++j) {
                int count = Dirs.Select(dir => (i + dir.Item1, j + dir.Item2)).Sum(Sum);
                if (board[i][j] > 0) {
                    if (count < 2 || count > 3) {
                        board[i][j] = 2;
                    }
                } else {
                    if (count == 3) {
                        board[i][j] = 3;
                    }
                }
            }
        }
        for (int i = 0; i < m; ++i) {
            for (int j = 0; j < n; ++j) {
                board[i][j] %= 2;
            }
        }
    }
}

